docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,62 @@
+The purpose of this file is to describe the functional and key architecture of
+this project. It is spiced with some rational throughout.
+
+# Slots & Tokens
+
+This library shall provide multiple tokens for the user(s). Any given slot can
+only contain one token at any given time. In case of smartcards the slot is a
+reader whilst the token is the actual smartcard. Since this project won't
+implement a separate tool for "virtual swapping" of tokens, each token is
+associated with its own slot. Thus, the library offers a set of slots with a
+token inside each of these.
+
+In order to allow the creation of new tokens using the pkcs11 means, this
+library presents one more slot/token pair than tokens currently exist. The
+additional token is a blank token that can be initialized using the C_InitToken
+function.
+(Note: This is not implemented yet)
+
+# Objects
+
+Each token can contain an arbitrary amount of keys and other object. (Note:
+other objects are not implemented yet). The keys are TPM-based keys, i.e. they
+are only known to the TPM and also the key operations are executed inside the
+TPM. The other objects are encrypted using hybrid encryption with the outer
+key being sealed to the TPM.
+
+Note: Since the TPM's object are stored (encrypted) external to the TPM on the
+HDD, it is not possible to delete them, if your attacker model allows access to
+the HDD.
```

Comment:
I'm not sure what this paragraph is trying to say. I think it's warning that if you delete the HDD-stored objects, they'll be lost? And that if you allow attackers to access your file system, they might be able to delete them?

If so, I think this could be more clearly worded as something like:

> Note: TPMs have very limited memory. To support an arbitrary number of objects, objects are stored (encrypted) in the file system and loaded/unloaded into the TPM as needed. If objects are deleted from the file system, they may be lost.

Reply:
It also means that if an attacker ever had read access to this location, when you delete it, it doesn't matter, because they have a copy of the object.